### PR TITLE
Update serial communication and clean up duplicated & outdated info

### PR DESCRIPTION
**Pull request by LDong-Arm**
This replaces #1317.

This PR is created to

replace third party serial software with the simple mbed sterm command
use printf() (instead of serial classes) for debug printing
clarify that buffered/unbuffered serial classes can capture keyboard inputs
remove/refactor duplicated information in multiple documents
The following files are refactored:

serial_communication.md - This is the guide for serial communication (mbed sterm, ...) with updated information.
quick-start-cli-debug.md, quick-start-online-debug.md - In the quick start guides, replace third-party serial software with mbed sterm.
debug_intro.md - As a guide for exporting project to IDE, mbed sterm does not really belong there.
debug_with_printf.md - This is now a pure printf guide that does not duplicate info in serial communication.
Note: minimal-printf is linked in debug_with_printf.md as before. Also, it's now the default printf implementation (ARMmbed/mbed-os#12233) so we shouldn't need to reiterate much in my opinion

### CODE REVIEW

**Comment by LDong-Arm** [comment]
Hi @iriark01 I created this to replace #1317 so we can preview the code. I also addressed some comments from there, please review.

Continuing our discussions, shall we remove debugging & exporting to IDEs from quick start?

**Comment by iriark01** [comment]
I think we should, yes. If the content is needed, it should be in the debugging area.

It would be a different matter if we had specific debugging info just for the quick start (for example, if Blinky had some commented-out broken lines that people could uncomment so we could walk them through debugging). But since it's standard printf() info we should just point to a good, single-source-of-truth page on printf().

**Comment by LDong-Arm** [comment]
@iriark01 I just removed the debugging pages in my last commit. The existing "Further reading" page of the quick start already links to debugging & serial communication docs, so I'm not adding extra links.

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
In addition to printing messages, Mbed OS applications can also read keyboard inputs from the user using the [BufferedSerial](../apis/bufferedserial.html) and [UnbufferedSerial](../apis/unbufferedserial.html) classes.
```

File: docs/tools/debug/debug_with_printf.md
Comment (iriark01) [inline docs/tools/debug/debug_with_printf.md]:
```suggestion
An easy way to inspect what your application is doing is to augment your application with `printf()` statements. In Mbed OS, you can use a serial connection to send feedback from your development board back to your computer.
```

File: docs/tools/debug/debug_with_printf.md
Comment (iriark01) [inline docs/tools/debug/debug_with_printf.md]:
```suggestion
For a tutorial on setting up serial communication, see [Board to PC communication over USB](../program-setup/serial-communication.html).
```

**Review by iriark01 — CHANGES_REQUESTED** [review]
Minor edits. I'm also building the branch so I can check that everything renders correctly, but I'll fix rendering issues without going through engineering approval again.

**Comment by LDong-Arm** [comment]
It looks like Travis failed because this needs a rebase to have the json fix (#1324):
```
Validating json ./docs/bare_metal/using_bare_metal.md:57 ...
Expecting property name: line 5 column 13 (char 90)
The command "./check_tools/find_bad_code_snippets.sh" exited with 1.
```

**Comment by iriark01** [comment]
But it will be fine once I merge it; the fix is already on "development", right?

**Comment by LDong-Arm** [comment]
> But it will be fine once I merge it; the fix is already on "development", right?

True

**Comment by iriark01** [comment]
So I think we just need Evelyne's review

**Review by iriark01 — APPROVED** [review]

File: docs/program-setup/serial/serial_communication.md
Comment (evedon) [inline docs/program-setup/serial/serial_communication.md]: It's probably worth linking to https://github.com/ARMmbed/mbed-os-tools/blob/master/packages/mbed-ls/README.md

**Comment by LDong-Arm** [comment]
@iriark01 There were a number of merge conflicts since the `development` branch changed. I tried to rebase this PR myself but I'm not sure if I did it correctly. Could you double check, especially the links/redirect, etc.?

(If I messed things up, please revert to 20eb660fb8f3b296a0aae5b2d62bd5a7ecf3e3e8, the hash before the rebase.)

Thanks.

File: docs/program-setup/serial/serial_communication.md
Comment (evedon) [inline docs/program-setup/serial/serial_communication.md]: The most common usage is auto-opening a serial terminal after compilation so I would put that section first and describe the sterm command subsequently.

File: docs/program-setup/serial/serial_communication.md
Comment (evedon) [inline docs/program-setup/serial/serial_communication.md]: You could also show usage of `-m detect`

**Comment by iriark01** [comment]
That commit looks fine.

File: docs/tools/debug/debugging_mbed_os_apps.md
Comment (evedon) [inline docs/tools/debug/debugging_mbed_os_apps.md:1]: Comment for line 5: this sentence should be removed or a link to debug_with_printf.md should be provided

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [reply docs/program-setup/serial/serial_communication.md]: I cannot find any page for `mbedls` on Mbed OS's documentation site. The info in the link seems outdated as it says "Development moved", and the steps to install this command don't apply anymore since `mbedls` is now included in Mbed CLI suite.

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [reply docs/program-setup/serial/serial_communication.md]: Good point, I just reordered and rephrased to introduce auto-opening after compilation first.

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [reply docs/program-setup/serial/serial_communication.md]: It's a nice tip. But `-m detect` seems like an undocumented feature (not shown in `mbed compile --help`), so I'm not sure if we should suggest it.

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [inline docs/program-setup/serial/serial_communication.md]: I've removed this because now we introduce `mbed compile --flash` method first, which doesn't require a reset to see "Hello World!" in our example.

File: docs/tools/debug/debugging_mbed_os_apps.md
Comment (LDong-Arm) [reply docs/tools/debug/debugging_mbed_os_apps.md:1]: Added links to printf and serial communication.

**Comment by LDong-Arm** [comment]
Hi @evedon, @iriark01, I just addressed latest comments from @evedon, please review the last three commits. Thanks.

**Review by evedon — APPROVED** [review]

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
You can use one command to compile an application, flash it onto a board and open a serial terminal:
```

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
When the board is flashed, it starts printing to the terminal (`Hello World!`, in this example).
```

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]: Note spans don't work for multiple lines, so I'll change this

Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
<div style="background-color:#F3F3F3; text-align:left; vertical-align: middle; padding:15px 30px;">  <b>Note:</b> </div>

```

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
<br/>- If your application uses a baud rate other than 9600, specify it with `-b <BAUDRATE>` in the command above.
```

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
<br/>- This method only works for a single board. To work with multiple boards, open a serial terminal manually as described below.</div>
```

Comment (iriark01) [reply docs/program-setup/serial/serial_communication.md]: I removed the bit about the target because it implied you could make this work for multiple boards of different targets, and we don't explain that.

Comment (iriark01) [reply docs/program-setup/serial/serial_communication.md]: As a user this will annoy me. You send me to "Manually opening a serial terminal" which then sends me to "Additional options" which explains how to connect multiple boards. Is there a way we can make this smoother? Since the two things explained in "Additional options" are how to use multiple boards and how to specify a baud rate (which you explained above), I think we can maybe we can do away with that bit. The `--help` bit can then have its own title or just be mentioned as a tip (we don't have to show output for `--help`).

**Review by iriark01 — CHANGES_REQUESTED** [review]

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [reply docs/program-setup/serial/serial_communication.md]: I agree we drop the "Additional options" section and replace it with a tip which is more compact.

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
<br/>1. Run `mbedls` to find the port of the board you want to use.
```

File: docs/program-setup/serial/serial_communication.md
Comment (iriark01) [inline docs/program-setup/serial/serial_communication.md]:
```suggestion
<br/>2. Run `mbed sterm -p <PORT>`.
```

Comment (iriark01) [reply docs/program-setup/serial/serial_communication.md]: (Once it's in a DIV it's HTML, not Markdown, so it doesn't auto-number)

**Review by iriark01 — APPROVED** [review]
Just some tidying of the note - a DIV makes it HTML instead of Markdown so it needs more manual handling

File: docs/program-setup/serial/serial_communication.md
Comment (LDong-Arm) [inline docs/program-setup/serial/serial_communication.md]: Between `<div>` and `</div>` formatting doesn't seem to work - `<BAUDRATE>` isn't shown, for example.

**Comment by LDong-Arm** [comment]
@iriark01 Thanks, the format looks better now

**Comment by iriark01** [comment]
Let's merge it and I'll clean it up after, because it's a bit funny in the preview. Is that okay?

**Comment by LDong-Arm** [comment]
Sure, sounds good to me

**Comment by iriark01** [comment]
Thank you very much for this PR!

**Comment by LDong-Arm** [comment]
Thanks for your help too!

**Pull request merged**
